### ISSUE TITLE
segfault with nested functions (?) or on return

### ISSUE THREAD

**Issue by jgoldfar** [JuliaLang/julia#9222]
I am using Julia for an application in PDE-constrained optimization, and have run into an issue I haven't seen before, giving a segfault when trying to return (apparently) from a function call. I only mention the nested functions because in the optimization setup, I construct functionals using passed data, and then pass *those* functionals to my optimization routine. Not sure if that is creating some issue behind the scenes. 

Based on the optimization trace, all of the iterations in the solver are working correctly, but then something breaks. The embedded PDE solver and optimization routine have passing unit tests, so I'm (pretty) sure the problem is not with their separate functionality, but somehow at the interface between them.

The gist with the backtrace is here https://gist.github.com/jgoldfar/b311086329af529fff42. The same issue appeared in version 0.3.0 I've been using for a while, but I figured I'd update to check if it resolved somehow. I'd be happy to look more into what is causing the issue on my own if someone could point me in the right direction!

**Comment by hayd** [comment]
Jeff's example segfaults in the repl for me on master with 3.7.1.

**Comment by tkelman** [comment]
what platform? we're not testing osx on travis right now because it's too slow, so hard to say whether #14687 would have triggered whatever you're seeing

**Comment by yuyichao** [comment]
Which platform is this and what's the error message?

**Comment by hayd** [comment]
That's OSX

```
julia> versioninfo()
Julia Version 0.5.0-dev+2172
Commit 3309d89 (2016-01-17 05:12 UTC)
Platform Info:
  System: Darwin (x86_64-apple-darwin14.5.0)
  CPU: Intel(R) Core(TM) i7-4650U CPU @ 1.70GHz
  WORD_SIZE: 64
  BLAS: libopenblas (USE64BITINT DYNAMIC_ARCH NO_AFFINITY Haswell)
  LAPACK: libopenblas64_
  LIBM: libopenlibm
  LLVM: libLLVM-3.7.1

julia> # paste in Jeff's example from above
[1]    3422 segmentation fault  ~/projects/julia/julia
```

**Comment by vtjnash** [comment]
this might be another instance of #9770, but i suspect it may actually be type-inference limit-tuple-type that is misbehaving. I think i just ran into a more likely cause of this bug a few days ago.

**Comment by vtjnash** [comment]
fixed by #15300, this would be a good case to add as an additional inference test before closing
